### PR TITLE
[ActiveStorage] Prevent `AS::Preview#processed` to generate a variant for an empty transformation

### PR DESCRIPTION
**Pull request by chaadow**
Followup #50044 cc @jonathanhefner 

### Motivation / Background

if an empty hash is passed to a preview call (`blob.preview({})`) We go through the original preview instead of regenerating a variation based on the original preview image which would result in a performance penalty.

### Detail

Calls to `#url` or `#key` would now use the original `blob.preview_image` instead of generating a variant that has an empty transformation.

==> Resulting in saving both a `MiniMagic/vips` call. + a network/IO call to the storage service

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Changes that are unrelated should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [x] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

### CODE REVIEW

File: activestorage/test/models/preview_test.rb [hunk 1]
```diff
@@ -70,6 +70,50 @@ class ActiveStorage::PreviewTest < ActiveSupport::TestCase
     assert_changes -> { !!preview.image.variant(transformations)&.send(:processed?) }, to: true do
       preview.processed
     end
+
+    image = read_image( preview.image.variant(transformations))
```
Comment (chaadow) [inline activestorage/test/models/preview_test.rb:74]: @jonathanhefner i've added a test to followup on your PR here.

also below,  i've added basic tests for all of `#download` `#key` and `#url` since they were missing and they're public API.

File: activestorage/test/models/preview_test.rb [hunk 2]
```diff
@@ -70,6 +70,61 @@ class ActiveStorage::PreviewTest < ActiveSupport::TestCase
     assert_changes -> { !!preview.image.variant(transformations)&.send(:processed?) }, to: true do
       preview.processed
     end
+
+    image = read_image(preview.image.variant(transformations))
+    assert_equal 216, image.width
+    assert_equal 280, image.height
```
Comment (jonathanhefner) [inline activestorage/test/models/preview_test.rb:76]: I don't think these assertions are necessary for this particular test.  This test only cares about whether the variant was processed.  Tests in `variant_test.rb` should check whether `Variant#processed` works as expected.

File: activestorage/test/models/preview_test.rb [hunk 3]
```diff
@@ -70,6 +70,61 @@ class ActiveStorage::PreviewTest < ActiveSupport::TestCase
...
+
+  test "#key does not generate a variant when provided an empty transformation" do
+    blob = create_file_blob(filename: "report.pdf", content_type: "application/pdf")
+    empty_transformation = {}
+    preview = blob.preview(empty_transformation)
+
+    assert_no_changes -> { !!preview.image.variant(empty_transformation)&.send(:processed?) } do
+      preview.processed.key
+    end
+  end
```
Comment (jonathanhefner) [inline activestorage/test/models/preview_test.rb:107]: I think we can combine these tests (and the `"#url does not generate a variant when provided an empty transformation"` test further below) into a single `"previewing with empty transformations does not generate a variant"` test.

File: activestorage/test/models/preview_test.rb [hunk 4]
```diff
@@ -70,6 +70,61 @@ class ActiveStorage::PreviewTest < ActiveSupport::TestCase
...
+
+    image = read_image(preview.image.variant(transformations))
+    assert_equal 216, image.width
+    assert_equal 280, image.height
+  end
+
+  test "#processed does not generate a variant when provided an empty transformation" do
+    blob = create_file_blob(filename: "report.pdf", content_type: "application/pdf")
+    empty_transformation = {}
+    preview = blob.preview(empty_transformation)
```
Comment (jonathanhefner) [inline activestorage/test/models/preview_test.rb:82]:
```suggestion
    preview = blob.preview({})
```

File: activestorage/test/models/preview_test.rb [hunk 5]
```diff
@@ -70,6 +70,61 @@ class ActiveStorage::PreviewTest < ActiveSupport::TestCase
...
+
+    assert_no_changes -> { !!preview.image.variant(empty_transformation)&.send(:processed?) } do
+      preview.processed
+    end
```
Comment (jonathanhefner) [inline activestorage/test/models/preview_test.rb:86]: Instead of `assert_no_changes`, let's just assert that there are no variant records after all is done:

```suggestion
    preview.processed
    # etc

    assert_empty preview.image.variant_records
```

File: activestorage/test/models/preview_test.rb [hunk 6]
```diff
@@ -70,6 +70,61 @@ class ActiveStorage::PreviewTest < ActiveSupport::TestCase
...
+  end
+
+  test "#key does not generate a variant when provided an empty transformation" do
+    blob = create_file_blob(filename: "report.pdf", content_type: "application/pdf")
+    empty_transformation = {}
+    preview = blob.preview(empty_transformation)
+
+    assert_no_changes -> { !!preview.image.variant(empty_transformation)&.send(:processed?) } do
+      preview.processed.key
```
Comment (jonathanhefner) [inline activestorage/test/models/preview_test.rb:105]: It might be good to also assert that `preview.key` is equal to `blob.preview_image.key`.

File: activestorage/test/models/preview_test.rb [hunk 7]
```diff
@@ -70,6 +70,61 @@ class ActiveStorage::PreviewTest < ActiveSupport::TestCase
...
+    end
+  end
+
+  test "#url does not generate a variant when provided an empty transformation" do
+    blob = create_file_blob(filename: "report.pdf", content_type: "application/pdf")
+    empty_transformation = {}
+    preview = blob.preview(empty_transformation)
+
+    assert_no_changes -> { !!preview.image.variant(empty_transformation)&.send(:processed?) } do
+      preview.processed.url
```
Comment (jonathanhefner) [inline activestorage/test/models/preview_test.rb:126]: It might be good to also assert that the `preview.url` is equal to `blob.preview_image.url`.

File: activestorage/test/models/preview_test.rb [hunk 8]
```diff
@@ -70,6 +70,61 @@ class ActiveStorage::PreviewTest < ActiveSupport::TestCase
...
+    preview = blob.preview(empty_transformation)
+
+    assert_no_changes -> { !!preview.image.variant(empty_transformation)&.send(:processed?) } do
+      preview.processed.key
+    end
+  end
+
+  test "#key #download and #url raise `UnprocessedError` when preview is not processed" do
+    blob = create_file_blob(filename: "report.pdf", content_type: "application/pdf")
+    preview = blob.preview({})
```
Comment (jonathanhefner) [inline activestorage/test/models/preview_test.rb:111]: Since this behavior isn't specific to empty transformations, and specifying non-empty transformations is the common case, let's do that here:

```suggestion
  test "methods raise UnprocessedError when preview is not processed" do
    blob = create_file_blob(filename: "report.pdf", content_type: "application/pdf")
    preview = blob.preview(resize_to_limit: [640, 280])
```

File: activestorage/CHANGELOG.md [hunk 9]
```diff
@@ -1,3 +1,9 @@
+*   Prevent `ActiveStorage::Blob#preview` to generate a variant if an empty variation is passed.
+    calls to `#url`, `#key` or `#download` will now use the original preview
```
Comment (jonathanhefner) [inline activestorage/CHANGELOG.md:2]:
```suggestion
    Calls to `#url`, `#key` or `#download` will now use the original preview
```

File: activestorage/app/models/active_storage/preview.rb [hunk 10]
```diff
@@ -105,14 +105,17 @@ def process
     end
 
     def variant
-      image.variant(variation).processed
+      image.variant(variation)
     end
 
-
```
[the remark this answers is not in the document]
Comment (chaadow) [reply activestorage/app/models/active_storage/preview.rb:111]: Yeah sorry it's my editor(vim) :/ i'll try to put it back

File: activestorage/test/models/preview_test.rb (hunk 7 above)
Comment (chaadow) [reply activestorage/test/models/preview_test.rb:126]: I agree

File: activestorage/test/models/preview_test.rb (hunk 8 above)
Comment (chaadow) [reply activestorage/test/models/preview_test.rb:111]: `methods raise ...` I find the "methods" very generic, I'm going to suggest a slight refactoring, and let me know what you think.

I agree with the non empty transformations 👍🏼

File: activestorage/test/models/preview_test.rb (hunk 2 above)
Comment (chaadow) [reply activestorage/test/models/preview_test.rb:76]: I agree

File: activestorage/app/models/active_storage/preview.rb [hunk 11]
```diff
@@ -63,7 +63,7 @@ def image
   # a stable URL that redirects to the URL returned by this method.
   def url(**options)
     if processed?
-      variant.url(**options)
+      empty_variation? ? image.url(**options) : variant.processed.url(**options)
```
[the remark this answers is not in the document]
Comment (chaadow) [reply activestorage/app/models/active_storage/preview.rb:66]: I love this

File: activestorage/app/models/active_storage/preview.rb (hunk 10 above)
[the remark this answers is not in the document]
Comment (chaadow) [reply activestorage/app/models/active_storage/preview.rb:111]: Done

File: activestorage/test/models/preview_test.rb [hunk 12]
```diff
@@ -4,6 +4,8 @@
 require "database/setup"
 
 class ActiveStorage::PreviewTest < ActiveSupport::TestCase
+  PREVIEW_METHODS = %i[key url download]
```
Comment (chaadow) [inline activestorage/test/models/preview_test.rb:7]: @jonathanhefner Since these methods are public method and public interface as well, I wanted to test them each individually even though they are similar.

I believe in the future, adding more public methods to AS:Preview would simply require adding the method under test to this constant.

Right now I believe I'm testing 2/3 of these methods. the 1/3 left is when provided a non blank transformation. I don't know if you think it's worth testing.

File: activestorage/test/models/preview_test.rb (hunk 3 above)
Comment (chaadow) [reply activestorage/test/models/preview_test.rb]: https://github.com/rails/rails/pull/50046#discussion_r1393143712

File: activestorage/test/models/preview_test.rb [hunk 13]
```diff
@@ -72,6 +74,41 @@ class ActiveStorage::PreviewTest < ActiveSupport::TestCase
...
+    preview = blob.preview({})
+
+    preview.processed
+
+    assert_empty preview.image.variant_records
+  end
+
+  PREVIEW_METHODS.each do |method_under_test|
+    test "##{method_under_test} does not generate a variant when provided an empty transformation" do
+      freeze_time do
```
Comment (chaadow) [inline activestorage/test/models/preview_test.rb:88]: I added `freeze_time` to handle random failure tests with calls to `#url`

File: activestorage/test/models/preview_test.rb (hunk 12 above)
Comment (chaadow) [reply activestorage/test/models/preview_test.rb:7]: I decided to add the missing third. And with this, I believe we have full coverage of all of the three preview methods.

**Comment by chaadow** [comment]
I've squashed and rebased with main.

File: activestorage/test/models/preview_test.rb [hunk 14]
```diff
@@ -72,6 +74,56 @@ class ActiveStorage::PreviewTest < ActiveSupport::TestCase
...
+  PREVIEW_METHODS.each do |method_under_test|
+    test "#{method_under_test} raises `UnprocessedError` when preview is not processed" do
+      blob = create_file_blob(filename: "report.pdf", content_type: "application/pdf")
+      preview = blob.preview(resize_to_limit: [640, 280])
+
+      assert_raises ActiveStorage::Preview::UnprocessedError do
+        preview.send(method_under_test)
+      end
+    end
+  end
```
Comment (jonathanhefner) [inline activestorage/test/models/preview_test.rb:125]: Declaring tests in a loop like this can be a good strategy for exhaustive test coverage.  However, in this case, I think it adds unnecessary complexity.  It is also a leaky abstraction.  `PREVIEW_METHODS` does not accommodate methods that require arguments.  So if we were to add such methods in the future, we would need to rework this entire approach.

I believe we can cover all the behavior we're interested in with the following two tests:

```ruby
test "image-related methods raise UnprocessedError when preview is not processed" do
  blob = create_file_blob(filename: "report.pdf", content_type: "application/pdf")
  preview = blob.preview(resize_to_limit: [640, 280])

  assert_raises(ActiveStorage::Preview::UnprocessedError) { preview.url }
  assert_raises(ActiveStorage::Preview::UnprocessedError) { preview.key }
  assert_raises(ActiveStorage::Preview::UnprocessedError) { preview.download }
end

test "previewing with empty transformations does not generate a variant" do
  blob = create_file_blob(filename: "report.pdf", content_type: "application/pdf")
  preview = blob.preview({})

  preview.processed

  freeze_time { assert_equal blob.preview_image.url, preview.url }
  assert_equal blob.preview_image.key, preview.key
  assert_equal blob.preview_image.download, preview.download

  assert_empty preview.image.variant_records
end
```

Let's use those instead.

**Review by jonathanhefner — COMMENTED** [review]
I see that you recently pushed many documentation changes.  Would you mind splitting those out into separate PRs unless they are directly related to this change?  It makes things easier to review.  Also, keep in mind that this is an optimization for a corner case, so I don't think we need to extensively document it (if at all).

**Comment by chaadow** [comment]
> Would you mind splitting those out into separate PRs unless they are directly related to this change?

@jonathanhefner I changed all of `processed`, `url`, `key` and `download`, and they're publicly documented API, I thought I would update the documentation to reflect the new reality of these methods. Even though it's an "edge case", I thought documentation should elicit all possible cases.

Moreover, contextually speaking, since it's a followup from our past Pull request that we co-authored, in which we changed `#processed` but we forgot to update the documentation, (and I also re-changed this method again), I thought I might update to further help the readers with multiple examples that could potentially help them by adding more context.

I will revert and create another PR. 

Thanks again for the prompt and exhaustive review!

File: activestorage/test/models/preview_test.rb (hunk 14 above)
Comment (chaadow) [reply activestorage/test/models/preview_test.rb]: Done. thanks

**Comment by chaadow** [comment]
@jonathanhefner I've applied the last changes, squashed the commits, and rebased with main.

**Comment by jonathanhefner** [comment]
Thank you, @chaadow! :heavy_check_mark:

**Comment by nicowenterodt** [comment]
Thanks from me as well. Very useful 👏 @chaadow @jonathanhefner

File: activestorage/test/models/preview_test.rb [hunk 15]
```diff
@@ -72,6 +72,27 @@ class ActiveStorage::PreviewTest < ActiveSupport::TestCase
...
+    assert_raises(ActiveStorage::Preview::UnprocessedError) { preview.download }
+  end
+
+  test "previewing with empty transformations does not generate a variant" do
+    blob = create_file_blob(filename: "report.pdf", content_type: "application/pdf")
+    preview = blob.preview({})
+
+    preview.processed
+
+    freeze_time { assert_equal blob.preview_image.url, preview.url }
```
Comment (zzak) [inline activestorage/test/models/preview_test.rb:90]: @chaadow I've stumbled upon this now, why is this here? If I remove it the tests still pass for me.
